### PR TITLE
Introduce librustc_index crate

### LINKED ISSUE #50592: extract index and bit-set related types into their own crate
Right now they live in rustc_data_structures, but we can rely on crates now. It'd be nicer if they were over in a rust-lang crate, which would let us optimize, test, and otherwise develop them independently from rustc. (It'd also be useful in projects like [borrow-check](https://github.com/rust-lang-nursery/borrow-check)). As you can see in [this comment](https://github.com/rust-lang/rust/issues/50592#issuecomment-510907510), the current plan is to begin doing this "in tree" and only later decide whether to move out of tree.

### PR DESCRIPTION
**Pull request by csmoe**
Closes #50592

### CODE REVIEW

File: src/librustc_index/lib.rs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#![feature(allow_internal_unstable)]
+#![feature(unboxed_closures)]
+#![feature(test)]
+#![feature(fn_traits)]
+
+pub mod indexed_vec;
```
Comment (eddyb) [inline src/librustc_index/lib.rs]: Maybe this should just be named `vec`?

Comment (csmoe) [reply src/librustc_index/lib.rs]: okay, will address soon.

**Comment by csmoe** [comment]
@eddyb almost no changes here, https://github.com/rust-lang/rust/pull/64778/files#diff-1240c4024b5a96328bb0436c2344a020L847, but I cannot figure out where the error came from.

**Comment by eddyb** [comment]
Presumably `rustdoc` is using a different copy of the `Idx` trait?

File: src/librustdoc/lib.rs [hunk 2]
```diff
@@ -25,6 +25,7 @@ extern crate getopts;
 extern crate env_logger;
 extern crate rustc;
 extern crate rustc_data_structures;
+extern crate rustc_index;
```
Comment (csmoe) [inline src/librustdoc/lib.rs]: @eddyb  This works now, but previous 2018-edition-style(add rustc_index inside Cargo.toml then `rustc_index::xxx`) failed unexpectedly, did I miss something?

Comment (eddyb) [reply src/librustdoc/lib.rs]: Aah, if you add it to Cargo.toml it creates a copy - so make sure you only add it to Cargo.toml if rustc_data_structures is already in Cargo.toml.

**Comment by eddyb** [comment]
cc @nikomatsakis 

@bors r+

**Pull request merged**